src/python/pants/jvm/goals/coursier.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os.path
+from dataclasses import dataclass
+
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.console import Console
+from pants.engine.fs import (
+    EMPTY_DIGEST,
+    CreateDigest,
+    Digest,
+    DigestContents,
+    FileContent,
+    MergeDigests,
+    Snapshot,
+    Workspace,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule
+from pants.engine.target import Sources, Target, Targets
+from pants.jvm.resolve.coursier_fetch import CoursierResolvedLockfile, MavenRequirements
+from pants.jvm.target_types import JvmLockfileSources, MavenRequirementsField
+
+
+class CoursierResolveSubsystem(GoalSubsystem):
+    name = "coursier-resolve"
+    help = "Generate a lockfile by resolving Maven dependencies."
+
+
+class CoursierResolve(Goal):
+    subsystem_cls = CoursierResolveSubsystem
+
+
+@dataclass(frozen=True)
+class CoursierGenerateLockfileRequest:
+    """Regenerate a coursier_lockfile target's lockfile from its Maven requirements.
+
+    This allows the user to manually regenerate their lockfile.  This is done for a few reasons: to
+    generate the lockfile for the first time, to regenerate it because the input Maven requirements
+    have changed, or to regenerate it to check if the resolve has changed (e.g. due to newer
+    versions of dependencies being published).
+    """
+
+    target: Target
+
+
+@dataclass(frozen=True)
+class CoursierGenerateLockfileResult:
+    digest: Digest
+
+
+@rule
+async def coursier_generate_lockfile(
+    request: CoursierGenerateLockfileRequest,
+) -> CoursierGenerateLockfileResult:
+    resolved_lockfile = await Get(
+        CoursierResolvedLockfile,
+        MavenRequirements,
+        MavenRequirements.create_from_maven_coordinates_fields(
+            fields=(request.target[MavenRequirementsField],),
+        ),
+    )
+    resolved_lockfile_json: bytes = resolved_lockfile.to_json()
```

Comment:
Not a big deal, but preferred is to use `cast`: https://www.pantsbuild.org/docs/style-guide#prefer-cast-to-override-annotations

Reply:
I think I didn't need an annotation at all here: `to_json()` already explicitly returns` `bytes`